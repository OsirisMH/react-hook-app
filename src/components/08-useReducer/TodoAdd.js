import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = React.memo(({ handleAddTodo }) => {
    const [ { description }, handleInputChange, reset ] = useForm({
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };


        handleAddTodo( newTodo );
        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form
                className="d-grid gap-2"
                onSubmit={ handleSubmit }
            >
                <input
                    autoComplete="off"
                    className="form-control"
                    name="description"
                    placeholder="Aprender..."
                    type="text"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary btn-sm"
                >
                    Agregar
                </button>
            </form>   
        </>
    )
});