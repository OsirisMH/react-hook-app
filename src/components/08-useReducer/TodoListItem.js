import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = React.memo(({ todo, index, handleDelete, handleToggle }) => {
    return (
        <li
            key={ todo.id }
            className="list-group-item d-fleX justify-content-between align-items-start"
        >
            <p 
                className={`${ todo.done && 'completed' }`}
                onClick={ () => handleToggle(todo.id) }
            >
                { index + 1 }. &emsp; { todo.desc }
            </p>
            <button
                className="btn btn-danger"
                onClick={ () => handleDelete(todo.id) }
            >
                Borrar
            </button>
        </li>
    )
})


TodoListItem.propTypes = {
    todo: PropTypes.object,
    index: PropTypes.number,
    handleDelete: PropTypes.func,
    handleToggle: PropTypes.func,
}