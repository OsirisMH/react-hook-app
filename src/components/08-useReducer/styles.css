body{
    padding: 30px 5%;
}

#root > div > div > div.col-7 > ul > li > p{
    cursor: pointer;
}

.list-group-item{
    align-items: center;
    display: flex;
}

.completed{
    text-decoration: line-through;
    color: rgb(26, 105, 37);
}