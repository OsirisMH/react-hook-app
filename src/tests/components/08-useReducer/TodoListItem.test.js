import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";

describe('Testing on <TodoListItem />', () => {
    
    const todo = demoTodos[0];
    const index = 0;
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoListItem 
            todo = { todo }
            index = { index }
            handleDelete = { handleDelete }
            handleToggle = { handleToggle }
        />
    );

    test('should render correctly', () => {
        //snapshot
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('should call the handleDelete function', () => {
        //jest.fn()
        wrapper.find('button').simulate('click');
        expect( handleDelete ).toHaveBeenCalledWith( todo.id );

    });
    
    test('should call handleToggle function', () => {
        //jest.fn()
        wrapper.find('p').simulate('click');
        expect( handleToggle ).toHaveBeenCalledWith( todo.id );

    });
    
    test('should render the text correctly', () => {

        const todoText = wrapper.find('p').text().trim();
        expect( todoText ).toBe( `${index + 1}. \u2003 ${todo.desc}` );

    });
    
    test('should has completed class if TODO.done = true', () => {
        const todo = demoTodos[0];
        todo.done = true;

        const wrapper = shallow(
            <TodoListItem 
                todo = { todo }
            />
        );

        expect( wrapper.find('p').hasClass('completed') ).toBe( true );
    });
    
});
