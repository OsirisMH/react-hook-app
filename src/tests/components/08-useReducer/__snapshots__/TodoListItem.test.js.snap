// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <TodoListItem /> should render correctly 1`] = `
<li
  className="list-group-item d-fleX justify-content-between align-items-start"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . â€ƒ 
    Aprender Mongo
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
